sw/device/silicon_creator/lib/drivers/otp.c

```diff
@@ -0,0 +1,80 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+
+#include <stddef.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "otp_ctrl_regs.h"
+
+static rom_error_t otp_read_start(const otp_t *otp, uint32_t address) {
+  uint32_t status =
+      mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+  // Check if otp_ctrl is busy.
+  if (!bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT)) {
+    return kErrorOtpBusy;
+  }
+
+  // Issue read command.
+  mmio_region_write32(
+      otp->base_addr, OTP_CTRL_DIRECT_ACCESS_CMD_REG_OFFSET,
+      bitfield_bit32_write(0u, OTP_CTRL_DIRECT_ACCESS_CMD_RD_BIT, true));
+
+  // Wait for read to complete.
+  while (true) {
+    status = mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+    // TODO If the DAI is to be used in mask_rom add more robust error checking.
+    if (bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_ERROR_BIT)) {
+      return kErrorOtpUnknown;
+    }
+    if (bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT)) {
+      return kErrorOk;
+    }
+  }
+}
+
+uint32_t otp_read32(const otp_t *otp, uint32_t address) {
+  return mmio_region_read32(otp->base_addr,
+                            OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
+}
+uint64_t otp_read64(const otp_t *otp, uint32_t address) {
```

Comment:
nit: one line of whitespace here (I think `git clang-tidy` will do that for you).

Reply:
One would think, not sure why clang-format doesn't pick up on this...